bench/hamming/trigonometry.fpcore

```diff
@@ -0,0 +1,89 @@
+
+
+; 2sin involves the difference of two sine functions. According to the sum-to-product trigonometric identity,
+; the target for this benchmark can apply the identity (- (sin x) (sin y)) = (* 2 (cos (/ (+ x y) 2)) (sin (/ (- x y) 2)))
+(FPCore (x eps)
+  :name "2sin (example 3.3)"
+  :pre (and (<= -1e4 x) (<= x 1e4) (< (* 1e-16 (fabs x)) eps) (< eps (* (fabs x))))
+  :herbie-target
+  (* 2 (cos (+ x (/ eps 2))) (sin (/ eps 2)))
+ 
+  (- (sin (+ x eps)) (sin x)))
+
+
+; 2cos involves the difference of two cos functions. According to the sum-to-product trigonometric identity,
+; the target for this benchmark can apply the identity (- (cos x) (cos y)) = (* 2 (sin (/ (+ x y) 2)) (sin (/ (- x y) 2)))
+(FPCore (x eps)
+ :name "2cos (problem 3.3.5)"
+ :pre (and (<= -1e4 x) (<= x 1e4) (< (* 1e-16 (fabs x)) eps) (< eps (* (fabs x))))
+ :herbie-target
+ (* -2 (sin (+ x (/ eps 2))) (sin (/ eps 2)))
+
+ (- (cos (+ x eps)) (cos x)))
+
+
+; 2tan involves the difference of two tan functions. A slightly more intuitive proof, this involves
+; converting the two tan into (- (/ (sin (+ x eps) (cos (+ x eps)))) (/ (sin x) (cos x))). After cross
+; multiplying we get the massive formula (/ (- (* (sin (+ x eps)) (cos x)) (* (sin x) (cos (+ x eps)))) (* (cos (+ x eps)) (cos x)))
+; Fortunately, the numerator is just an expansion of the identity sin(a-b) where a = x+eps and b = x.
+; Hence, we can simply use that identity to craft the numerator and leaving the denominator as is.
+(FPCore (x eps)
+ :name "2tan (problem 3.3.2)"
+ :pre (and (<= -1e4 x) (<= x 1e4) (< (* 1e-16 (fabs x)) eps) (< eps (* (fabs x))))
+ :herbie-target
+ (/ (sin eps) (* (cos x) (cos (+ x eps))))
+
+ (- (tan (+ x eps)) (tan x)))
+
+; This is directly taken from the Hamming textbook but this is a direct identity of the trig half identity
+; for tan, and we completely remove any trace of a subtraction using this, fulfilling our goal.
+; As a result, we can query any value of x from -infinity to +infinity as we eliminated any cause
+; of the error using this rewrite.
+(FPCore (x)
+ :name "tanhf (example 3.4)"
+ :pre (and (<= (- INFINITY) x) (<= x INFINITY))
+ :herbie-target
+ (tan (/ x 2))
+
+ (/ (- 1 (cos x)) (sin x)))
+
+
+; Noting the difference in the variable name, N is actually a large N. Hence, we are changing our precondition
+; definition by including all values of N till 1e+100. 
+; For the rewrite, we use the trigonometric identity, (- (atan x) (atan y)) = (atan (/ (- x y) (+ 1 (* x y))))
+(FPCore (N)
+ :name "2atan (example 3.5)"
+ :pre (and (> N 1) (< N 1e+100))
+ :herbie-target
+ (atan (/ 1 (+ 1 (* N (+ N 1)))))
+
+ (- (atan (+ N 1)) (atan N)))
+
+
+; This is a slightly unique rewrite because it involves two different trigonometric functions that don't have
+; a direct rewrite. Moreover, Hamming gave this as an exercise with no real hint on how to approach this.
+; Hence, we resort to the Taylor series of this identity where the rewrite for all values for (<= (fabs x) 1)
```

Comment:
Ok but when x is close to 1, this Taylor expansion is bad... For example, at x = 1 you get only about 32 bits right, which ain't bad but still not ideal. Maybe we don't have better ideas but I bet adding a branch would improve things. I bet around +/- 0.2 is the right place for the branch.